wcpredictor/src/bpl_interface.py

```diff
@@ -190,20 +212,62 @@ def get_fixture_goal_probabilities(
         goals = np.arange(0, max_goals + 1)
         probs = {}
         simulated_scores = []
-        # if fixture_teams is not passed, we just predict for all games in self.fixtures by default
+        # if fixture_teams not passed, predict for all games in self.fixtures by default
         if fixture_teams is None:
             fixture_teams = self.get_fixture_teams()
         Team_1, Team_2 = zip(*fixture_teams)
         Team_1_conference = [self.confed_dict[team] for team in Team_1]
         Team_2_conference = [self.confed_dict[team] for team in Team_2]
         venue = np.ones(len(Team_1))
         for i in range(len(fixture_teams)):
-            home_team_goal_prob = self.model.predict_score_n_proba(
-                goals, Team_1[i], Team_2[i], Team_1_conference[i], Team_2_conference[i], home=False, neutral_venue=venue[i]
-            )
-            away_team_goal_prob = self.model.predict_score_n_proba(
-                goals, Team_2[i], Team_1[i], Team_2_conference[i], Team_1_conference[i], home=False, neutral_venue=venue[i]
-            )
+            if isinstance(self.model, NeutralDixonColesMatchPredictorWC):
+                home_team_goal_prob = self.model.predict_score_n_proba(
+                    goals,
+                    Team_1[i],
+                    Team_2[i],
+                    Team_1_conference[i],
+                    Team_2_conference[i],
+                    home=False,
+                    neutral_venue=venue[i],
+                )
+                away_team_goal_prob = self.model.predict_score_n_proba(
+                    goals,
+                    Team_2[i],
+                    Team_1[i],
+                    Team_2_conference[i],
+                    Team_1_conference[i],
+                    home=False,
+                    neutral_venue=venue[i],
+                )
+            elif isinstance(self.model, NeutralDixonColesMatchPredictor):
+                home_team_goal_prob = self.model.predict_score_n_proba(
+                    goals,
+                    Team_1[i],
+                    Team_2[i],
+                    home=False,
```

Comment:
yep all your comments are correct (they're the same on main, though, they just appear in the diff here as the `confederation-priors` branch isn't up to date with main). I think these functions in WCPred are assuming it's only dealing with world cup matches. I'm not using them for what I'm doing but will fix them up once I get a hyperparameter scan launched

Reply:
ah I see you beat me to it and already fixed it 😁

Reply:
I think they would be wrong though because it would have the teams the wrong way round, so would be simulating the away team vs away team even if it was neutral?

Reply:
in fact, should it be
```
home_team_goal_prob = self.model.predict_score_n_proba(
                goals, Team_1[i], Team_2[i], Team_1_conference[i], Team_2_conference[i], home=True, neutral_venue=venue[i]
            )
            away_team_goal_prob = self.model.predict_score_n_proba(
                goals, Team_1[i], Team_2[i], Team_1_conference[i], Team_2_conference[i], home=False, neutral_venue=venue[i]
            )
```?

Reply:
In `.predict_score_n_proba()`, we have:
```
probs = (
            self.predict_score_proba(
                team_rep,
                opponent_rep,
                team_conf_rep,
                opponent_conf_rep,
                n_rep,
                x_rep,
                neutral_venue_rep,
            )
            if home
            else self.predict_score_proba(
                opponent_rep,
                team_rep,
                opponent_conf_rep,
                team_conf_rep,
                x_rep,
                n_rep,
                neutral_venue_rep,
            )
        ).reshape(MAX_GOALS + 1, len(n))
```
the line:
```
home_team_goal_prob = self.model.predict_score_n_proba(
    goals,
    Team_1[i],
    Team_2[i],
    Team_1_conference[i],
    Team_2_conference[i],
    home=False,
    neutral_venue=venue[i],
)
```
with `home=True`, we're predicting how many goals `Team_1[i]` scores against `Team_2[i]` (which is fine).

But the line:
```
away_team_goal_prob = self.model.predict_score_n_proba(
    goals,
    Team_2[i],
    Team_1[i],
    Team_2_conference[i],
    Team_1_conference[i],
    home=False,
    neutral_venue=venue[i],
)
```
with `home=False` means that we switch around `Team_2[i]` and `Team_1[i]` and we end up predicting how many goals `Team_1[i]` scores against `Team_2[i]`.

So we really should have
```
away_team_goal_prob = self.model.predict_score_n_proba(
    goals,
    Team_1[i],
    Team_2[i],
    Team_1_conference[i],
    Team_2_conference[i],
    home=False,
    neutral_venue=venue[i],
)
```
instead, right?

Reply:
it seems like this is everywhere (even in `bpl_framework.py` in AIrsenal)

Reply:
I believe your first fix is correct, rather than the above. The two above give the probability of Team_1 scoring  n goals against Team_2 when Team_1 is at home (first line) and when Team_1 is away (2nd line). The way the `predict_score_n_proba` functions are defined is a bit confusing when shown like this (that function specifically always returns the probabilities for the first team passed, Team_1 in this case, only, i.e. the order of the teams is specifying which team we're interested in, not which team is playing at home)

I guess these will have been impacting the WC simulations, but I do think only for Qatar (as they're the only team who home/away will make a difference for, assuming neutral venue has been set correctly).

Reply:
ah okay, I see that it switches around `x_rep` and `n_rep` too so I agree that it's right now. that's a relief 😅

Reply:
If there's still anything unclear about how `self.model.predict_score_n_proba` works happy to chat about it tomorrow!

Reply:
after some sleep and a couple of `print` statements, I think this makes sense now. the `x_rep` and `n_rep` get swapped over as well so you end up getting the probabilities for the away team. Confusing though 😅

but I agree what we have is correct now